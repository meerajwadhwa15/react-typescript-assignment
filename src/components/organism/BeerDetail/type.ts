import { Action, ActionCreator } from 'redux'
import { StringLiteral } from '@babel/types'
import ErrorMessage from 'components/atoms/ErrorMessage'

export type itemType = {
  id: number
  image_url: string
  name: string
  tagline: string
  volume: {
    value: number
    unit: string
  }
  description: string
  food_pairing: [string]
  first_brewed: string
}

export type ComponentProps = {
  match: {
    params: {
      id: string
    }
  }
}

export type ComponentState = {}

export interface ReducerState {
  item: itemType
  isFetching: boolean
  errorMessage: string
}

export interface DispatchProps {
  fetchBeerDetailAction: (id: number) => void
}

export type ReducerInitialeState = {
  item: {}
  isFetching: boolean
}

export type BeerDetailReducerType = {
  BeerDetailReducer: {
    item: itemType
    isFetching: boolean
    errorMessage: string
  }
}

export type ReducerAction = {
  type: string
  item: itemType[]
  isFetching: boolean
  message: string
}

export interface FetchActionType extends Action {
  type: 'FETCH_BEER_DETAIL'
}

export interface FetchSuccessActionType extends Action {
  type: 'FETCH_BEER_DETAIL_SUCCESS'
  item: itemType
}

export interface SetFetch {
  type: 'SET_FETCHING_DETAIL'
  isFetching: boolean
}

export type FetchErrorActionType = {
  type: string
  message: string
}
