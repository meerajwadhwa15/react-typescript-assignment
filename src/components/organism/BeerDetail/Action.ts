import { ActionCreator, Dispatch, AnyAction } from 'redux'
import { AxiosResponse, AxiosError } from 'axios'
import {
  FetchActionType,
  FetchSuccessActionType,
  itemType,
  SetFetch,
  FetchErrorActionType,
} from './type'
import { ThunkAction, ThunkDispatch } from 'redux-thunk'
import Fetch, { ServerProps } from './../../../libs/fetch'
import { ERROR_MESSAGE } from './../../../config/message'
import { BEER_BY_ID } from './../../../config/api'

export const isFetching = (isFetching: boolean): SetFetch => {
  return { type: 'SET_FETCHING_DETAIL', isFetching }
}

export const FetchBeerDetail = (id: number) => {
  return async (dispatch: ThunkDispatch<{}, {}, AnyAction>): Promise<void> => {
    const showLoader: SetFetch = isFetching(true)
    dispatch(showLoader)
    const serverProps: ServerProps = {
      url: BEER_BY_ID.replace('{ID}', String(id)),
    }

    return Fetch(serverProps)
      .then((response: AxiosResponse) => {
        const fetchBeersSuccess: FetchSuccessActionType = {
          type: 'FETCH_BEER_DETAIL_SUCCESS',
          item: response.data,
        }
        dispatch(fetchBeersSuccess)
      })
      .catch((err: AxiosError) => {
        const fetchBeersSuccess: FetchErrorActionType = {
          type: 'FETCH_BEER_DETAIL_ERROR',
          message: ERROR_MESSAGE,
        }
        dispatch(fetchBeersSuccess)
      })
      .finally(() => {
        const hideLoader: SetFetch = isFetching(false)
        dispatch(hideLoader)
      })
  }
}
