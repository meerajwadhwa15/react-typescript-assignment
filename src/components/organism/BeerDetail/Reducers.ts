import { Reducer } from 'redux'
import {
  ReducerState,
  ReducerAction,
  itemType,
  ReducerInitialeState,
} from './type'

export const initialState: ReducerInitialeState = {
  item: {},
  isFetching: false,
}

const BeerDetailReducer: Reducer<ReducerInitialeState, ReducerAction> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case 'SET_FETCHING_DETAIL':
      return { ...state, isFetching: action.isFetching }
    case 'FETCH_BEER_DETAIL_SUCCESS':
      return { ...state, item: action.item[0] }
    case 'FETCH_BEER_DETAIL_ERROR':
      return { ...state, errorMessage: action.message }
    default:
      return state
  }
}

export default BeerDetailReducer
