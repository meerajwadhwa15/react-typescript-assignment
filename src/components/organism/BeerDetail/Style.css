.item-detail {
  color: #2c2c2c;
  display: flex;
  line-height: 22px;
  margin-bottom: 20px;
  padding: 15px;
}

.item-detail h2 {
  font-size: 16px;
  font-weight: bold;
}

.item-detail .label {
  color: #2c2c2c;
  font-size: 14px;
  font-weight: bold;
}

.item-detail .detail {
  padding-left: 15px;
}

.item-detail .detail > div,
.item-detail .detail > p,
.item-detail .detail > h2 {
  margin-bottom: 10px;
}
