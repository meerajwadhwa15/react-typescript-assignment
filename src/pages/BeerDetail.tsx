import React, { FunctionComponent } from 'react'
import Layout from './../components/templates/Layout'
import BeerDetail from './../components/organism/BeerDetail'

type Props = {
  match: {
    params: {
      id: string
    }
  }
}

const BeerDetailPage: FunctionComponent<Props> = (props) => (
  <Layout>
    <BeerDetail {...props} />
  </Layout>
)

export default BeerDetailPage
