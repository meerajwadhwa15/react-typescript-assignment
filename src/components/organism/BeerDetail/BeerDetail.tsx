import React, { FunctionComponent, ReactNode, Component, Fragment } from 'react'
import { Dispatch } from 'redux'
import { ThunkDispatch } from 'redux-thunk'
import {
  ComponentProps,
  ReducerState,
  BeerDetailReducerType,
  DispatchProps,
  ComponentState,
} from './type'
import Img from './../../atoms/Img'
import { connect } from 'react-redux'
import { FetchBeerDetail } from './Action'
import Loader from './../../molecules/Loader'
import ErrorMessage from './../../atoms/ErrorMessage'
import NoRecordFound from './../../atoms/NoRecordFound'
import FoodPairing from './../../molecules/FoodPairing'
import './Style.css'

export const mapStateToProps = ({
  BeerDetailReducer: { item, isFetching, errorMessage },
}: BeerDetailReducerType): ReducerState => ({
  item,
  isFetching,
  errorMessage,
})

export const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  fetchBeerDetailAction: (id: number) => {
    dispatch(FetchBeerDetail(id))
  },
})

export type Props = ReducerState & DispatchProps & ComponentProps

export class BeerDetail extends Component<Props, ComponentState> {
  componentDidMount() {
    const id = parseInt(this.props.match.params.id, 10)
    if (!this.props.item || !this.props.item.id || this.props.item.id !== id) {
      this.props.fetchBeerDetailAction(id)
    }
  }

  render() {
    const { item, isFetching, errorMessage } = this.props

    return (
      <div>
        {isFetching ? (
          <Loader />
        ) : errorMessage ? (
          <ErrorMessage>{errorMessage}</ErrorMessage>
        ) : item.id ? (
          <div className="item-detail">
            <div>
              <Img src={item.image_url} height="300" alt={item.tagline} />
            </div>
            <div className="detail">
              <h2>{item.name}</h2>
              <div>{item.tagline}</div>
              <div>
                {item.volume ? item.volume.value : null}{' '}
                {item.volume ? item.volume.unit : null}
              </div>
              <p>{item.description}</p>
              <FoodPairing foodPairing={item.food_pairing} />
              <div>
                <strong className="label">Brew Date:</strong>&nbsp;
                {item.first_brewed}
              </div>
            </div>
          </div>
        ) : (
          <NoRecordFound>No Record Found.</NoRecordFound>
        )}
      </div>
    )
  }
}

// Now let's connect our component!
export default connect<
  ReducerState,
  DispatchProps,
  ComponentProps,
  BeerDetailReducerType
>(
  mapStateToProps,
  mapDispatchToProps
)(BeerDetail)
