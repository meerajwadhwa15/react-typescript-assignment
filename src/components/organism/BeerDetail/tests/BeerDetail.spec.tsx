import React from 'react'
import { shallow } from 'enzyme'
import { BeerDetail, mapStateToProps, Props } from './../BeerDetail'
import { create } from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'

const sampleLoadingMockData: Props = {
  fetchBeerDetailAction: jest.fn(),
  item: {},
  isFetching: true,
  match: {
    params: {
      id: '1',
    },
  },
}

const sampleListMockData: Props = {
  fetchBeerDetailAction: jest.fn(),
  item: {
    id: 1,
    image_url: 'https://images.punkapi.com/v2/keg.png',
    name: 'Buzz',
    tagline: 'A Real Bitter Experience.',
    volume: {
      value: 20,
      unit: 'litres',
    },
  },
  isFetching: false,
  match: {
    params: {
      id: '1',
    },
  },
}

const ComponentWithLoading = () => (
  <BrowserRouter>
    <BeerDetail {...sampleListMockData} />
  </BrowserRouter>
)

const ComponentWithList = () => (
  <BrowserRouter>
    <BeerDetail {...sampleLoadingMockData} />
  </BrowserRouter>
)

describe('BeerList Enzyme testing', () => {
  it('renders with Loading without crashing', () => {
    shallow(<ComponentWithLoading />)
  })

  it('renders with beerList without crashing', () => {
    shallow(<ComponentWithList />)
  })

  test('Matches the snapshot with beer List', () => {
    const component = create(<ComponentWithList />)
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('Matches the snapshot without beer list', () => {
    const component = create(<ComponentWithLoading />)
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('Should call fetchBeersAction ', () => {
    const spy = jest.spyOn(sampleLoadingMockData, 'fetchBeerDetailAction')
    const component = create(<BeerDetail {...sampleLoadingMockData} />)
    const instance = component.getInstance()
    instance.componentDidMount()
    expect(spy).toBeCalled()
  })

  test('Should test mapStateToProps', () => {
    expect(
      mapStateToProps({
        BeerDetailReducer: {
          item: {},
          isFetching: false,
        },
      })
    ).toEqual({
      item: {},
      isFetching: false,
    })
  })
})
