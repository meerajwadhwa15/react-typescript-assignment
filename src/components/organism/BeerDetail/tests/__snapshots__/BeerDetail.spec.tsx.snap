// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerList Enzyme testing Matches the snapshot with beer List 1`] = `
<div>
  <div
    className="loader"
  >
    Loading...
  </div>
</div>
`;

exports[`BeerList Enzyme testing Matches the snapshot without beer list 1`] = `
<div>
  <div
    className="item-detail"
  >
    <div>
      <img
        alt="A Real Bitter Experience."
        height="300"
        src="https://images.punkapi.com/v2/keg.png"
      />
    </div>
    <div
      className="detail"
    >
      <h2>
        Buzz
      </h2>
      <div>
        A Real Bitter Experience.
      </div>
      <div>
        20
         
        litres
      </div>
      <p />
      <div />
      <div>
        <strong
          className="label"
        >
          Brew Date:
        </strong>
        Â 
      </div>
    </div>
  </div>
</div>
`;
